Basics

Vseslav0 opened the pull request.

Vseslav0 committed:
Add files via upload

Vseslav0 committed:
Create README.md

Vseslav0 committed:
Delete labs/basics/basics directory

Vseslav0 committed:
Rename labs/basics/BUILD to labs/basics/basics1/BUILD

Vseslav0 committed:
Rename labs/basics/main.cpp to labs/basics/basics1/main.cpp

Vseslav0 committed:
Rename labs/basics/mainwindow.cpp to labs/basics/basics1/mainwindow.cpp

Vseslav0 committed:
Rename labs/basics/mainwindow.h to labs/basics/basics1/mainwindow.h

Vseslav0 committed:
Rename labs/basics/ui_mainwindow.h to labs/basics/basics1/ui_mainwindow.h

Vseslav0 committed:
Add files via upload

Vseslav0 committed:
Rename labs/basics/BUILD to labs/basics/basics2/BUILD

Vseslav0 committed:
Rename labs/basics/README.md to labs/basics/basics2/README.md

Vseslav0 committed:
Rename labs/basics/main.cpp to labs/basics/basics2/main.cpp

Vseslav0 committed:
Rename labs/basics/mainwindow.cpp to labs/basics/basics2/mainwindow.cpp

Vseslav0 committed:
Rename labs/basics/mainwindow.h to labs/basics/basics2/mainwindow.h

Vseslav0 committed:
Rename labs/basics/ui_mainwindow.h to labs/basics/basics2/ui_mainwindow.h

Skir0 commented:
## Задание 1
 ### Виджеты
 * count: QSpinBox: позволяет ввести общее количество всех билетов (ok)
* view: QListвWidget или QTableWidget: показывает общий прогресс в повторении билетов. (ok)
* question-view: QButtonGroup, QGroupBox или что-то подобное. Содержит виджеты, позволяющие просматривать и изменять информацию о текущем билете: номер, название, редактирование названия, изменение статуса (реализовано как всплывающее окно)
> **БАГ**
> * При выборе последующего билета, окно прошлого билета не закрывается. Тем самым при повторном выборе билета, окно не отображается (оно находится за главным окном). И изменение билетов возможно сразу во всех окнах, что по-моему некорректно. В итоге получается такая картина:
<img width="1128" alt="Screenshot 2025-03-29 at 11 01 56" src="https://github.com/user-attachments/assets/3a62efd1-e0f4-415a-b85b-986014509c23" /> 

> * У status bar я бы поменял названия ячеек с цветов на названия "выучил", "не выучил", "надо повторить" для интуитивного понимания. 

> * Дефолтный (цвет билетов при загрузке списка билетов) и белый цвет я бы сделал как один

* next-question: QPushButton, при нажатии на который выбирается следующий рандомный вопрос — только из вопросов с дефолтным либо желтым статусом. Во view соответствующая ячейка должна становиться активной, question-view должно отображать соответствующий билет.
> **БАГ**
> * Из-за окон, если билет был до этого выбран и он выпадает как рандомный, то он не отображается в question_view

> * "Зеленые билеты" могут выпадать, что не соответствует тз

 > * Пул рандомных билетов не соответствует пулу всех возможных билетов. Видимо из-за того, что когда появляются зеленые билеты, количество билетов в пуле сокращается и не может быть выбран условно последний билет, номер которого соответствует размеру нового пула + 1 (старому пулу)
 
* previous-question: QPushButton, показывающая вопрос, который был выбран до текущего. Во view и question-view должны проводиться изменения, аналогичные тем, что в next-question.
> **БАГ**
> * Некорректная логика работы
> _Советую реализовать логику этой кнопки через stack_

* total-progress: QProgressBar, обозначающий прогресс в повторении. Увеличивается, если билет из дефолтного становится желтым или зеленым (и уменьшается, если зеленый или желтый билет становится дефолтным).
* green-progress: как total-progress, но увеличивается, если желтый или дефолтный билет становится зеленым.
> **БАГ**
![Uploading Screenshot 2025-03-29 at 12.22.47.png…]()

 ### Доп. критерии
*  Layout и Resize (ok)
*  Дизайн (простой, но у меня не лучше, поэтому не осуждаю) (ok)

### Некоторые комментарии
1. Дабл клик не может с желтого сделать зеленый
2. Я бы поменял цвета, так как желтый с белым смотрится очень плохо (сам желтый бы оставил, просто выбрал другой оттенок) + белый и белый - такое себе)

Skir0 commented:
## Задание 2
**Баги**
У меня лично была вот такая штука при запуске
![Screenshot 2025-03-30 at 14 57 45](https://github.com/user-attachments/assets/e7a6c045-4bb0-4778-854c-7a5fb74e3d54)
я заменил строку на 
```
std::random_device rd;
std::mt19937 g(rd());
std::shuffle(elements.begin(), elements.end(), g);
```

**Функционал**
 Описаный функционал соответствует действительности: все работает. Пару комментариев:
 * в mode 1 я бы добавил проверку на пустую строку, так как при пустой строке выдает yes/no
 * в mode 2 можно добавить больше сепараторов последовательности (кроме " " добавить ",", ";" и т.д.). Также я бы как-то обрабатывал пробелы (из пробелов идущих подряд только один является разделителем, что не очень красиво при получении результата)

kexbr commented:
# Задание 1
Мой вердикт будет схожим с предыдущим проверяющим. 
## Виджеты
- [x] count: QSpinBox: позволяет ввести общее количество всех билетов
- [x] view: QListWidget или QTableWidget: показывает общий прогресс в повторении билетов.
- [x] question-view: QButtonGroup, QGroupBox или что-то подобное. Содержит виджеты, позволяющие просматривать и изменять информацию о текущем билете: номер, название, редактирование названия, изменение статуса.
```Предыдущий ревьювер указал, что выпадает слишком много окон, которые не закрываются. В тз ничего не было сказано по этому поводу (что оно должно закрываться), поэтому это не баг скорее, а фича.```
- [x] next-question: QPushButton, при нажатии на который выбирается следующий рандомный вопрос — только из вопросов с дефолтным либо желтым статусом. Во view соответствующая ячейка должна становиться активной, question-view должно отображать соответствующий билет.
```Тут да, возникают проблемы со следующими билетами. Не может попасться последний билет. Но у меня не было такого, чтобы выпадал билет, который помечен зелёным цветом.```
- [x] previous-question: QPushButton, показывающая вопрос, который был выбран до текущего. Во view и question-view должны проводиться изменения, аналогичные тем, что в next-question.
```Предыдущая кнопка 'зацикливается' и выдаёт только два последних билета. Как сказал предыдущий, реализуйте с помощью стека поддержку последних билетов, чтобы можно было легко получать из него предыдущий билет.```
- [x] total-progress: QProgressBar, обозначающий прогресс в повторении. Увеличивается, если билет из дефолтного становится желтым или зеленым (и уменьшается, если зеленый или желтый билет становится дефолтным).
- [x] green-progress: как total-progress, но увеличивается, если желтый или дефолтный билет становится зеленым.

Также, как и указал предыдущий, жёлтый билет при дабл клике не может стать зелёным, нужно это делать через окошко. Необходимо пофиксить

### Доп. критерии
- [x] Layout и Resize
- [x] Дизайн
### Мнение
В целом всё окей, все виджеты на месте и работают, но есть пару багов, которые, как по мне, достаточно легко фиксятся. Можно ещё немного поработать над названиями, как и сказал предыдущий. Доп фичей нет, но вся база выполнена. Ставлю 9 кружек пива из 10 🍺🍺🍺

kexbr commented:
# Задача 2
Прикольное приложение, как по мне очень полезное. У меня зачастую стоит выбор что-то делать либо не делать, и я очень часто решаю через [random.org](https://random.org) делать это или нет. Виджетов хватает, всё функционирует и коннектов 6, как и требуется. Есть нетривиальные виджеты. Было бы классно, если бы сепараторов было побольше (к примеру ",", ";"). А так довольно креативно и интересно. 10 судьбоносных предсказаний из 10🔮🔮🔮

Vseslav0 committed:
Update mainwindow.cpp

Фикс бага, при котором не мог выпасть последний билет, добавлена очередь билетов

Vseslav0 committed:
Update mainwindow.h

pkasila commented on labs/basics/basics2/mainwindow.cpp:
```diff
@@ -0,0 +1,150 @@
+
+void RandomField::Random() {
+    auto elements = parent_->Sequence()->text().split(" ");
+    std::random_shuffle(elements.begin(), elements.end());
```
```suggestion
    std::random_device rd;
    std::mt19937 g(rd());
    std::shuffle(elements.begin(), elements.end(), g);
```

https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4190.htm
